In [2]:
import pandas as pd

In [3]:
df=pd.read_csv(r'C:\Users\SYED  MOHAMMED GHOUS\Downloads\Messy_Employee_dataset.csv')
df

,Employee_ID,First_Name,Last_Name,Age,Department_Region,Status,Join_Date,Salary,Email,Phone,Performance_Score,Remote_Work
0,EMP1000,Bob,Davis,25.0,DevOps-California,Active,04-02-2021,59767.65,bob.davis@example.com,-1651623197,Average,True
1,EMP1001,Bob,Brown,NaN,Finance-Texas,Active,07-10-2020,65304.66,bob.brown@example.com,-1898471390,Excellent,True
2,EMP1002,Alice,Jones,NaN,Admin-Nevada,Pending,12-07-2023,88145.90,alice.jones@example.com,-5596363211,Good,True
3,EMP1003,Eva,Davis,25.0,Admin-Nevada,Inactive,11/27/2021,69450.99,eva.davis@example.com,-3476490784,Good,True
4,EMP1004,Frank,Williams,25.0,Cloud Tech-Florida,Active,01-05-2022,109324.61,frank.williams@example.com,-1586734256,Poor,False
...,...,...,...,...,...,...,...,...,...,...,...,...
1015,EMP2015,David,Miller,30.0,HR-California,Active,8/19/2023,NaN,david.miller@example.com,-3546212759,Good,True
1016,EMP2016,David,Johnson,30.0,Cloud Tech-Texas,Inactive,11-07-2021,100215.06,david.johnson@example.com,-2508261122,Good,True
1017,EMP2017,Charlie,Williams,40.0,Finance-New York,Active,10-04-2023,114587.11,charlie.williams@example.com,-1261632487,Average,False
1018,EMP2018,Alice,Garcia,30.0,HR-Florida,Inactive,12/16/2024,71318.79,alice.garcia@example.com,-8995729892,Good,True


In [4]:
df.shape

(1020, 12)

In [5]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1020 entries, 0 to 1019
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Employee_ID        1020 non-null   str    
 1   First_Name         1020 non-null   str    
 2   Last_Name          1020 non-null   str    
 3   Age                809 non-null    float64
 4   Department_Region  1020 non-null   str    
 5   Status             1020 non-null   str    
 6   Join_Date          1020 non-null   str    
 7   Salary             996 non-null    float64
 8   Email              1020 non-null   str    
 9   Phone              1020 non-null   int64  
 10  Performance_Score  1020 non-null   str    
 11  Remote_Work        1020 non-null   bool   
dtypes: bool(1), float64(2), int64(1), str(8)
memory usage: 88.8 KB


In [6]:
df.isnull().sum()

Employee_ID            0
First_Name             0
Last_Name              0
Age                  211
Department_Region      0
Status                 0
Join_Date              0
Salary                24
Email                  0
Phone                  0
Performance_Score      0
Remote_Work            0
dtype: int64

In [7]:
df.duplicated().sum()

np.int64(0)

In [8]:
df.dtypes

Employee_ID              str
First_Name               str
Last_Name                str
Age                  float64
Department_Region        str
Status                   str
Join_Date                str
Salary               float64
Email                    str
Phone                  int64
Performance_Score        str
Remote_Work             bool
dtype: object

In [9]:
df['Join_Date'].head(10)

0    04-02-2021
1    07-10-2020
2    12-07-2023
3    11/27/2021
4    01-05-2022
5    06-10-2020
6    04-03-2020
7     7/17/2022
8    12-08-2023
9    08-04-2022
Name: Join_Date, dtype: str

In [10]:
df['Join_Date'] = pd.to_datetime(df['Join_Date'], errors='coerce')

In [11]:
df['Join_Date'].dtype

dtype('<M8[us]')

In [12]:
df['Join_Date'].isnull().sum()

np.int64(620)

In [13]:
df['Phone'].head(10)

0   -1651623197
1   -1898471390
2   -5596363211
3   -3476490784
4   -1586734256
5   -5409003485
6   -4518376063
7   -4134327559
8   -4177656123
9   -8156985699
Name: Phone, dtype: int64

In [14]:
df['Phone'].tail(10)

1010   -8307482072
1011   -6511893939
1012   -6557820189
1013   -4322302332
1014   -2470739200
1015   -3546212759
1016   -2508261122
1017   -1261632487
1018   -8995729892
1019   -7629745492
Name: Phone, dtype: int64

In [15]:
df['Phone'].isnull().sum()

np.int64(0)

In [16]:
(df['Phone'] < 0).sum()

np.int64(1020)

In [17]:
df['Phone'].min(), df['Phone'].max()

(np.int64(-9994973202), np.int64(-3896086))

In [18]:
df['Phone'].abs().astype(str).str.len().value_counts()

Phone
10    928
9      78
8      12
7       2
Name: count, dtype: int64

In [19]:
df[df['Phone'].abs().astype(str).str.len()<10]

,Employee_ID,First_Name,Last_Name,Age,Department_Region,Status,Join_Date,Salary,Email,Phone,Performance_Score,Remote_Work
43,EMP1043,Bob,Garcia,NaN,Admin-California,Inactive,NaT,64744.08,bob.garcia@example.com,-164597793,Excellent,False
45,EMP1045,Alice,Brown,25.0,HR-New York,Pending,NaT,61310.74,alice.brown@example.com,-61027768,Average,True
47,EMP1047,Alice,Davis,35.0,Sales-Illinois,Active,NaT,58921.19,alice.davis@example.com,-310303722,Average,True
58,EMP1058,David,Williams,35.0,Finance-Texas,Pending,2021-11-07,93620.88,david.williams@example.com,-54522800,Good,False
68,EMP1068,Bob,Brown,40.0,Admin-Florida,Pending,NaT,119152.47,bob.brown@example.com,-846684549,Poor,True
...,...,...,...,...,...,...,...,...,...,...,...,...
946,EMP1946,Charlie,Jones,25.0,Admin-Illinois,Inactive,2022-02-02,104516.55,charlie.jones@example.com,-652775843,Excellent,False
948,EMP1948,Bob,Miller,40.0,Finance-Texas,Active,2024-08-03,NaN,bob.miller@example.com,-109121721,Excellent,True
986,EMP1986,Grace,Brown,30.0,Finance-Nevada,Active,NaT,109523.40,grace.brown@example.com,-849378305,Poor,True
1002,EMP2002,Bob,Smith,35.0,Finance-Illinois,Pending,NaT,63049.22,bob.smith@example.com,-586950731,Good,True


In [20]:
df['Phone'].abs().astype(str).str.len().value_counts().sort_index()

Phone
7       2
8      12
9      78
10    928
Name: count, dtype: int64

In [21]:
df['Phone'] = df['Phone'].abs()

In [22]:
df['Phone'] = df['Phone'].astype(str)

In [23]:
df['Phone'].head()

0    1651623197
1    1898471390
2    5596363211
3    3476490784
4    1586734256
Name: Phone, dtype: str

In [24]:
df['Phone'].astype(str)

0       1651623197
1       1898471390
2       5596363211
3       3476490784
4       1586734256
           ...    
1015    3546212759
1016    2508261122
1017    1261632487
1018    8995729892
1019    7629745492
Name: Phone, Length: 1020, dtype: str

In [25]:
df['Performance_Score'].value_counts()

Performance_Score
Good         270
Average      267
Excellent    267
Poor         216
Name: count, dtype: int64

In [26]:
df['Performance_Score'].isnull().sum()

np.int64(0)

In [27]:
df['Status'].value_counts()

Status
Pending     356
Active      352
Inactive    312
Name: count, dtype: int64

In [28]:
df['Status'].isnull().sum()

np.int64(0)

In [29]:
df['Department_Region'].value_counts()

Department_Region
HR-Florida               41
DevOps-California        35
Sales-Nevada             35
Admin-Nevada             34
Admin-California         34
DevOps-Florida           34
DevOps-New York          33
HR-New York              33
Sales-Florida            33
DevOps-Illinois          33
Finance-Illinois         33
Finance-California       32
Sales-California         31
Sales-Illinois           30
Finance-Texas            29
Admin-Illinois           29
Finance-Nevada           29
Cloud Tech-Texas         29
Cloud Tech-California    29
Cloud Tech-Florida       28
DevOps-Texas             27
DevOps-Nevada            27
Sales-New York           26
HR-California            26
Admin-Florida            25
Cloud Tech-New York      24
Finance-Florida          24
HR-Nevada                24
HR-Illinois              24
Sales-Texas              23
HR-Texas                 23
Finance-New York         23
Admin-Texas              22
Admin-New York           22
Cloud Tech-Nevada        20
Cl

In [30]:
df['Department_Region'].str.split('-').str.len().value_counts()

Department_Region
2    1020
Name: count, dtype: int64

In [31]:
df['Age'].describe()

count    809.000000
mean      32.484549
std        5.656860
min       25.000000
25%       25.000000
50%       30.000000
75%       40.000000
max       40.000000
Name: Age, dtype: float64

In [32]:
df['Age'].value_counts().sort_index()

Age
25.0    206
30.0    205
35.0    188
40.0    210
Name: count, dtype: int64

In [33]:
df['Age'] = df['Age'].fillna(df['Age'].median())

In [34]:
df['Age'].isnull().sum()    


np.int64(0)

In [35]:
df['Age'].describe()

count    1020.000000
mean       31.970588
std         5.136901
min        25.000000
25%        30.000000
50%        30.000000
75%        35.000000
max        40.000000
Name: Age, dtype: float64

In [36]:
df['Age'].value_counts()

Age
30.0    416
40.0    210
25.0    206
35.0    188
Name: count, dtype: int64

In [37]:
df['Salary'].describe()

count       996.000000
mean      85155.056396
std       19873.727918
min       50047.320000
25%       68392.487500
50%       85547.870000
75%      100974.027500
max      119971.650000
Name: Salary, dtype: float64

In [38]:
df['Salary'].info()

<class 'pandas.Series'>
RangeIndex: 1020 entries, 0 to 1019
Series name: Salary
Non-Null Count  Dtype  
--------------  -----  
996 non-null    float64
dtypes: float64(1)
memory usage: 8.1 KB


In [39]:
df['Salary'].value_counts().sort_index()    

Salary
50047.32     1
50060.73     1
50110.66     1
50153.79     1
50173.12     1
            ..
119586.11    1
119764.20    1
119801.30    1
119890.35    1
119971.65    1
Name: count, Length: 978, dtype: int64

In [40]:
df['Salary'] = df['Salary'].fillna(df['Salary'].median())

In [41]:
df['Salary'].isnull().sum()

np.int64(0)

In [42]:
df['Email'].isnull().sum()

np.int64(0)

In [43]:
df['Email'].duplicated().sum()

np.int64(956)

In [44]:
df['Remote_Work'].value_counts()

Remote_Work
True     513
False    507
Name: count, dtype: int64

In [45]:
df['Remote_Work'].isnull().sum()

np.int64(0)

In [46]:
df['Employee_ID'].isnull().sum()

np.int64(0)

In [47]:
df['Employee_ID'].duplicated().sum()

np.int64(0)

In [48]:
df['First_Name'].isnull().sum()
df['Last_Name'].isnull().sum()

np.int64(0)

In [49]:
df['Phone_Length'] = df['Phone'].str.len()

In [50]:
df['Phone_Length'].value_counts().sort_index()

Phone_Length
7       2
8      12
9      78
10    928
Name: count, dtype: int64

In [51]:
df.isnull().sum()

Employee_ID            0
First_Name             0
Last_Name              0
Age                    0
Department_Region      0
Status                 0
Join_Date            620
Salary                 0
Email                  0
Phone                  0
Performance_Score      0
Remote_Work            0
Phone_Length           0
dtype: int64

In [52]:
df.drop('Phone_Length', axis=1, inplace=True)

In [53]:
df.shape

(1020, 12)

In [54]:
df.duplicated().sum()

np.int64(0)

In [55]:
df[['Department', 'Region']] = df['Department_Region'].str.split('-', expand=True)
df

,Employee_ID,First_Name,Last_Name,Age,Department_Region,Status,Join_Date,Salary,Email,Phone,Performance_Score,Remote_Work,Department,Region
0,EMP1000,Bob,Davis,25.0,DevOps-California,Active,2021-04-02,59767.65,bob.davis@example.com,1651623197,Average,True,DevOps,California
1,EMP1001,Bob,Brown,30.0,Finance-Texas,Active,2020-07-10,65304.66,bob.brown@example.com,1898471390,Excellent,True,Finance,Texas
2,EMP1002,Alice,Jones,30.0,Admin-Nevada,Pending,2023-12-07,88145.90,alice.jones@example.com,5596363211,Good,True,Admin,Nevada
3,EMP1003,Eva,Davis,25.0,Admin-Nevada,Inactive,NaT,69450.99,eva.davis@example.com,3476490784,Good,True,Admin,Nevada
4,EMP1004,Frank,Williams,25.0,Cloud Tech-Florida,Active,2022-01-05,109324.61,frank.williams@example.com,1586734256,Poor,False,Cloud Tech,Florida
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1015,EMP2015,David,Miller,30.0,HR-California,Active,NaT,85547.87,david.miller@example.com,3546212759,Good,True,HR,California
1016,EMP2016,David,Johnson,30.0,Cloud Tech-Texas,Inactive,2021-11-07,100215.06,david.johnson@example.com,2508261122,Good,True,Cloud Tech,Texas
1017,EMP2017,Charlie,Williams,40.0,Finance-New York,Active,2023-10-04,114587.11,charlie.williams@example.com,1261632487,Average,False,Finance,New York
1018,EMP2018,Alice,Garcia,30.0,HR-Florida,Inactive,NaT,71318.79,alice.garcia@example.com,8995729892,Good,True,HR,Florida


In [56]:
df['Department'].value_counts()

Department
DevOps        189
Sales         178
HR            171
Finance       170
Admin         166
Cloud Tech    146
Name: count, dtype: int64

In [57]:
df['Region'].value_counts()

Region
California    187
Florida       185
Nevada        169
Illinois      165
New York      161
Texas         153
Name: count, dtype: int64

In [58]:
df['Status'].value_counts()

Status
Pending     356
Active      352
Inactive    312
Name: count, dtype: int64

In [59]:
df['Performance_Score'].value_counts()

Performance_Score
Good         270
Average      267
Excellent    267
Poor         216
Name: count, dtype: int64

In [60]:
df.groupby('Department')['Salary'].mean().sort_values(ascending=False)

Department
Sales         86873.948258
DevOps        86003.638624
HR            85475.379474
Admin         85202.644096
Cloud Tech    84944.434658
Finance       82279.519353
Name: Salary, dtype: float64

In [61]:
df['Salary'].describe()

count      1020.000000
mean      85164.299069
std       19638.385740
min       50047.320000
25%       68811.232500
50%       85547.870000
75%      100372.662500
max      119971.650000
Name: Salary, dtype: float64

In [62]:
df['Age'].describe()

count    1020.000000
mean       31.970588
std         5.136901
min        25.000000
25%        30.000000
50%        30.000000
75%        35.000000
max        40.000000
Name: Age, dtype: float64

In [63]:
df[['Age', 'Salary']].corr()

,Age,Salary
Age,1.000000,0.076761
Salary,0.076761,1.000000


In [64]:
df.groupby('Remote_Work')['Salary'].mean().sort_values(ascending=False)

Remote_Work
False    85835.187219
True     84501.257563
Name: Salary, dtype: float64

In [65]:
df.groupby('Status')['Salary'].mean().sort_values(ascending=False)

Status
Pending     85733.620927
Active      85139.967727
Inactive    84542.138974
Name: Salary, dtype: float64

In [66]:
df.groupby(['Department', 'Region'])['Salary'].mean().sort_values(ascending=False).head(10)

Department  Region    
Admin       Illinois      92050.898621
Cloud Tech  Florida       91715.487857
Sales       Florida       90582.736667
DevOps      Nevada        90001.822593
Admin       California    89564.108824
Finance     New York      89159.340435
Sales       California    88533.919677
HR          Texas         88007.280870
            California    87745.022692
            Illinois      87707.819167
Name: Salary, dtype: float64

In [67]:
df.groupby(['Department', 'Region'])['Salary'].mean().sort_values(ascending=False).tail(10)

Department  Region    
Finance     Texas         83832.979310
            Florida       83098.247917
DevOps      California    82772.097143
Admin       Texas         81871.540000
            Nevada        80641.359412
HR          Florida       80460.818293
Admin       New York      78822.267727
Cloud Tech  New York      78086.093333
Finance     Nevada        76398.523448
            Illinois      75707.702424
Name: Salary, dtype: float64

In [68]:
df

,Employee_ID,First_Name,Last_Name,Age,Department_Region,Status,Join_Date,Salary,Email,Phone,Performance_Score,Remote_Work,Department,Region
0,EMP1000,Bob,Davis,25.0,DevOps-California,Active,2021-04-02,59767.65,bob.davis@example.com,1651623197,Average,True,DevOps,California
1,EMP1001,Bob,Brown,30.0,Finance-Texas,Active,2020-07-10,65304.66,bob.brown@example.com,1898471390,Excellent,True,Finance,Texas
2,EMP1002,Alice,Jones,30.0,Admin-Nevada,Pending,2023-12-07,88145.90,alice.jones@example.com,5596363211,Good,True,Admin,Nevada
3,EMP1003,Eva,Davis,25.0,Admin-Nevada,Inactive,NaT,69450.99,eva.davis@example.com,3476490784,Good,True,Admin,Nevada
4,EMP1004,Frank,Williams,25.0,Cloud Tech-Florida,Active,2022-01-05,109324.61,frank.williams@example.com,1586734256,Poor,False,Cloud Tech,Florida
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1015,EMP2015,David,Miller,30.0,HR-California,Active,NaT,85547.87,david.miller@example.com,3546212759,Good,True,HR,California
1016,EMP2016,David,Johnson,30.0,Cloud Tech-Texas,Inactive,2021-11-07,100215.06,david.johnson@example.com,2508261122,Good,True,Cloud Tech,Texas
1017,EMP2017,Charlie,Williams,40.0,Finance-New York,Active,2023-10-04,114587.11,charlie.williams@example.com,1261632487,Average,False,Finance,New York
1018,EMP2018,Alice,Garcia,30.0,HR-Florida,Inactive,NaT,71318.79,alice.garcia@example.com,8995729892,Good,True,HR,Florida


In [69]:
df.isna().sum()

Employee_ID            0
First_Name             0
Last_Name              0
Age                    0
Department_Region      0
Status                 0
Join_Date            620
Salary                 0
Email                  0
Phone                  0
Performance_Score      0
Remote_Work            0
Department             0
Region                 0
dtype: int64

In [70]:
df.duplicated().sum()

np.int64(0)

In [71]:
df.dtypes

Employee_ID                     str
First_Name                      str
Last_Name                       str
Age                         float64
Department_Region               str
Status                          str
Join_Date            datetime64[us]
Salary                      float64
Email                           str
Phone                           str
Performance_Score               str
Remote_Work                    bool
Department                      str
Region                          str
dtype: object

In [72]:
df.duplicated().sum()

np.int64(0)

In [73]:
df['Phone'].head(5)

0    1651623197
1    1898471390
2    5596363211
3    3476490784
4    1586734256
Name: Phone, dtype: str

In [ ]:
df = df.drop(columns=['Department_Region'])

In [78]:
df['Status'].unique()
df['Performance_Score'].unique()
df['Remote_Work'].unique()

array([ True, False])

In [80]:
df['Status'].unique()
df['Performance_Score'].unique()

<StringArray>
['Average', 'Excellent', 'Good', 'Poor']
Length: 4, dtype: str

In [82]:
df['Performance_Score'].unique()

<StringArray>
['Average', 'Excellent', 'Good', 'Poor']
Length: 4, dtype: str

In [84]:
df['Status'].unique()

<StringArray>
['Active', 'Pending', 'Inactive']
Length: 3, dtype: str

In [87]:
Q1 = df['Salary'].quantile(0.25)
Q3 = df['Salary'].quantile(0.75)
IQR = Q3 - Q1
lower = Q1 - 1.5 * IQR
upper = Q3 + 1.5 * IQR
outliers = df[(df['Salary'] < lower) | (df['Salary'] > upper)]
print(f"Lower bound: {lower}, Upper bound: {upper}")
print(f"Number of outliers: {len(outliers)}")Q1 = df['Salary'].quantile(0.25)
Q3 = df['Salary'].quantile(0.75)
IQR = Q3 - Q1
lower = Q1 - 1.5 * IQR
upper = Q3 + 1.5 * IQR
outliers = df[(df['Salary'] < lower) | (df['Salary'] > upper)]
print(f"Lower bound: {lower}, Upper bound: {upper}")
print(f"Number of outliers: {len(outliers)}")

SyntaxError: invalid syntax (1435949756.py, line 8)

In [89]:
Q1 = df['Salary'].quantile(0.25)
Q3 = df['Salary'].quantile(0.75)
IQR = Q3 - Q1
lower = Q1 - 1.5 * IQR
upper = Q3 + 1.5 * IQR
outliers = df[(df['Salary'] < lower) | (df['Salary'] > upper)]
print(f"Lower bound: {lower}, Upper bound: {upper}")
print(f"Number of outliers: {len(outliers)}")

Lower bound: 21469.087499999987, Upper bound: 147714.80750000002
Number of outliers: 0


In [91]:
Q1 = df['Age'].quantile(0.25)
Q3 = df['Age'].quantile(0.75)
IQR = Q3 - Q1
lower = Q1 - 1.5 * IQR
upper = Q3 + 1.5 * IQR
outliers = df[(df['Age'] < lower) | (df['Age'] > upper)]
print(f"Lower bound: {lower}, Upper bound: {upper}")
print(f"Number of outliers: {len(outliers)}")

Lower bound: 22.5, Upper bound: 42.5
Number of outliers: 0


In [95]:
df.to_csv(r'C:\Users\SYED  MOHAMMED GHOUS\Downloads\Messy_Employee_dataset_cleaned.csv', index=False)



## Data Cleaning Summary — Before vs. After

| Metric | Before Cleaning | After Cleaning |
|---|---|---|
| Total rows | 1,020 | 1,020 |
| Total columns | 12 | 13 (Department_Region split into Department + Region) |
| Missing values — Age | 211 | 0 (filled with median) |
| Missing values — Salary | 24 | 0 (filled with median) |
| Duplicate rows | 0 | 0 |
| Join_Date dtype | object (string) | datetime64 |
| Phone dtype | int64 (negative values) | string |
| Salary outliers (IQR method) | Not checked | 0 found |
| Age outliers (IQR method) | Not checked | 0 found |
| Category formatting (Status, Performance_Score) | Unchecked | Verified clean — no inconsistencies |